controllers/collector_controller.go

```diff
@@ -92,95 +93,122 @@ func (r *CollectorReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
+func (r *CollectorReconciler) mutate(obj client.Object, co *fluentbitv1alpha2.Collector) controllerutil.MutateFn {
 	switch o := obj.(type) {
+	case *rbacv1.ClusterRole:
+		expected, _, _ := operator.MakeRBACObjects(co.Name, co.Namespace, "collector", co.Spec.RBACRules, co.Spec.ServiceAccountAnnotations)
+
+		return func() error {
+			o.Rules = expected.Rules
+			return nil
+		}
+	case *corev1.ServiceAccount:
+		_, expected, _ := operator.MakeRBACObjects(co.Name, co.Namespace, "collector", co.Spec.RBACRules, co.Spec.ServiceAccountAnnotations)
+
+		return func() error {
+			o.Annotations = expected.Annotations
+			if err := ctrl.SetControllerReference(co, o, r.Scheme); err != nil {
+				return err
+			}
+			return nil
+		}
+	case *rbacv1.ClusterRoleBinding:
+		_, _, expected := operator.MakeRBACObjects(co.Name, co.Namespace, "collector", co.Spec.RBACRules, co.Spec.ServiceAccountAnnotations)
+
+		return func() error {
+			o.RoleRef = expected.RoleRef
+			o.Subjects = expected.Subjects
+			return nil
+		}
 	case *appsv1.StatefulSet:
-		expected := operator.MakefbStatefulset(co)
+		expected := operator.MakefbStatefulset(*co)
 
 		return func() error {
 			o.Labels = expected.Labels
-			o.Annotations = expected.Annotations
 			o.Spec = expected.Spec
-			o.SetOwnerReferences(nil)
-			if err := ctrl.SetControllerReference(&co, o, r.Scheme); err != nil {
+			if err := ctrl.SetControllerReference(co, o, r.Scheme); err != nil {
 				return err
 			}
 			return nil
 		}
 	case *corev1.Service:
-		expected := operator.MakeCollectorService(co)
+		expected := operator.MakeCollectorService(*co)
 
 		return func() error {
 			o.Labels = expected.Labels
 			o.Spec.Selector = expected.Spec.Selector
 			o.Spec.Ports = expected.Spec.Ports
-			o.SetOwnerReferences(nil)
-			if err := ctrl.SetControllerReference(&co, o, r.Scheme); err != nil {
+			if err := ctrl.SetControllerReference(co, o, r.Scheme); err != nil {
 				return err
 			}
 			return nil
 		}
-
 	default:
 	}
 
 	return nil
 }
 
 func (r *CollectorReconciler) delete(ctx context.Context, co *fluentbitv1alpha2.Collector) error {
-	var sa corev1.ServiceAccount
+	sa := corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      co.Name,
+			Namespace: co.Namespace,
+		},
+	}
 	if err := r.Delete(ctx, &sa); err != nil && !errors.IsNotFound(err) {
 		return err
 	}
+	// TODO: clusterrole, clusterrolebinding
 
-	var svc corev1.Service
-	if err := r.Delete(ctx, &svc); err != nil && !errors.IsNotFound(err) {
+	sts := appsv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      co.Name,
+			Namespace: co.Namespace,
+		},
+	}
+	if err := r.Delete(ctx, &sts); err != nil && !errors.IsNotFound(err) {
 		return err
 	}
 
-	var sts appsv1.StatefulSet
-	if err := r.Delete(ctx, &sts); err != nil && !errors.IsNotFound(err) {
+	svc := corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      co.Name,
+			Namespace: co.Namespace,
```

Comment:
https://github.com/fluent/fluent-operator/blob/master/pkg/operator/collector-service.go#L35
The name of the service does not have to be `co.name`

Reply:
correct. i'll fix it.